chaos-mesh/chaos-mesh
# feat: support cgroup v2 for linux stress experiments

### What problem does this PR solve?

Support cgroup v2 driver when running experiments on linux (CPU & Mem). If the host/containers where chaosdaemon is running is in full cgroup v2 mode (unified), users will no longer see:
```
Failed to apply chaos: rpc error: code = Unknown desc = controller is not supported
```
When they run CPU or Memory stress experiments.

### What's changed and how it works?

Introduced a cgroupv2 manager to add newly created stress processes to the respective cgroup if containers based on what cgroup driver containers are using.

P.S. I did not re-order import statements myself, it must have been done by the format/lint/tidy goals in the Makefile when I ran make check or make test. I can revert these changes, if need be.

### Related changes

- [ ] Need to update `chaos-mesh/website`
- [ ] Need to update `Dashboard UI`
- Need to **cheery-pick to release branches**
  - [ ] release-2.1
  - [ ] release-2.0

### Checklist

Tests



  - Start kubernetes 1.21+ with cgroupfs and cgroup v2 driver.
  - Verify what cgroup driver you have on the node and in the container:

    ```bash
    # container
    $ kubectl -n test-dev exec -ti demo-c7c688b8-trr42 -- bash
    bash-4.2# mount| grep cgroup
    cgroup on /sys/fs/cgroup type cgroup2 (ro,nosuid,nodev,noexec,relatime,seclabel)

    # kubernetes node
    lvdkbw503 ~ # mount| grep cgroup
    cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,seclabel)
    ```

  - Schedule new CPU/Mem stress experiment and verify it's working by checking cgroup assigned to respective processes on the worker nodes.
  - Verify it on the node by using pids of the main process in the container and of the stress-ng/memstress process:

    ```bash
     lvdkbw503 ~ # ps -o cgroup 1476124
     CGROUP
     0::/kubepods/burstable/pod8731e44b-df00-41f9-9c50-
     1c6327a16b74/69604c2dd135683bf0758cf623c05c37f6fe61ab437fd2227b0f584e93eedbc0
     lvdkbw503 ~ # ps -o cgroup 2188035
     CGROUP
     0::/kubepods/burstable/pod8731e44b-df00-41f9-9c50-
     1c6327a16b74/69604c2dd135683bf0758cf623c05c37f6fe61ab437fd2227b0f584e93eedbc0
    ```

Side effects

N/A

### Release note 

```text
Please add a release note.

You can safely ignore this section if you don't think this PR needs a release note.
```

### DCO

If you find the DCO check fails, please run commands like below (Depends on the actual situations. For example, if the failed commit isn't the most recent) to fix it:

```shell
git commit --amend --signoff
git push --force
```
